File: src/main/java/com/naminhyeok/fantazzk/room/RoomResponse.java
package com.naminhyeok.fantazzk.room;

import java.util.Map;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

record RoomResponse(
    String code,
    String status,
    String mode,
    int teamCount,
    int teamSize,
    Integer budget,
    Integer minBidUnit,
    String draftOrderStrategy,
    String startReadiness,
    DraftOrderPreviewResponse draftOrderPreview,
    List<TeamLeaderResponse> teamLeaders,
    List<RoomPlayerResponse> players,
    List<RoomMemberResponse> members,
    RoomProgressResponse progress
) {
    static RoomResponse from(Room room) {
        List<RoomPlayer> players = room.getPlayers();
        Map<RoomPlayerId, RoomPlayer> playersById = players.stream()
            .collect(Collectors.toMap(RoomPlayer::getId, Function.identity()));
        return new RoomResponse(
            room.getCode(),
            room.getStatus().name(),
            room.getMode().name(),
            room.getTeamCount(),
            room.getTeamSize(),
            room.getBudget(),
            room.getMinBidUnit(),
            room.getDraftOrderStrategy() == null ? null : room.getDraftOrderStrategy().name(),
            room.getStartReadiness().name(),
            DraftOrderPreviewResponse.from(room),
            room.getLeaders().stream().map(TeamLeaderResponse::from).toList(),
            players.stream().map(RoomPlayerResponse::from).toList(),
            room.getMembers().stream()
                .map(member -> {
                    RoomPlayer player = playersById.get(member.playerId());
                    return RoomMemberResponse.from(member, player == null ? room.findPlayer(member.playerId()) : player);
                })
                .toList(),
            RoomProgressResponse.from(room)
        );
    }
}

File: team-building/api/src/main/kotlin/com/naminhyeok/fantazzk/teambuilding/dto/RoomResponse.kt
package com.naminhyeok.fantazzk.teambuilding.dto

import com.naminhyeok.fantazzk.teambuilding.room.Room
import com.naminhyeok.fantazzk.teambuilding.room.RoomStatus

data class RoomResponse(
    val code: String,
    val status: RoomStatus,
    val teamLeaders: List<TeamLeaderResponse>,
    val currentTarget: String?,
    val result: List<TeamResponse>?,
) {
    companion object {
        fun from(room: Room): RoomResponse =
            RoomResponse(
                code = room.code,
                status = room.status,
                teamLeaders =
                    room.teamLeaders.values.map {
                        TeamLeaderResponse(
                            id = it.id.value,
                            nickname = it.nickname,
                            remainingBudget = it.remainingBudget,
                            teamSize = it.team.size,
                        )
                    },
                currentTarget = room.playerPool.currentTarget()?.name,
                result =
                    room.result?.teams?.map { team ->
                        TeamResponse(
                            leaderNickname = team.leaderNickname,
                            members = team.members.map { it.name },
                        )
                    },
            )
    }
}

data class TeamLeaderResponse(
    val id: String,
    val nickname: String,
    val remainingBudget: Int?,
    val teamSize: Int,
)

data class TeamResponse(
    val leaderNickname: String,
    val members: List<String>,
)
